### PR TITLE
Fix EarlyStopping stop at fisrt epoch when patience=0 ; add auc to au…

### PR DESCRIPTION
**Pull request by DachuanZhao**
Fix issues :
[https://github.com/tensorflow/tensorflow/issues/49227](https://github.com/tensorflow/tensorflow/issues/49227)
[https://github.com/tensorflow/tensorflow/issues/44509](https://github.com/tensorflow/tensorflow/issues/44509)

### CODE REVIEW

File: keras/callbacks.py [hunk 1]
```diff
@@ -1777,7 +1777,7 @@ def __init__(self,
     elif mode == 'max':
       self.monitor_op = np.greater
     else:
-      if 'acc' in self.monitor:
+      if 'acc' in self.monitor or 'auc' in self.monitor:
```
[the remark this answers is not in the document]
Comment (DachuanZhao) [reply keras/callbacks.py]: OK , I have deleted it .

File: keras/callbacks.py [hunk 2]
```diff
@@ -1801,23 +1801,25 @@ def on_epoch_end(self, epoch, logs=None):
...
-    if self.wait >= self.patience:
-      self.stopped_epoch = epoch
-      self.model.stop_training = True
-      if self.restore_best_weights and self.best_weights is not None:
-        if self.verbose > 0:
-          print('Restoring model weights from the end of the best epoch.')
-        self.model.set_weights(self.best_weights)
+    else:
+      # Only judge the best after first epoch
+      self.wait += 1
```
Comment (fchollet) [inline keras/callbacks.py:1807]: I really don't see how incrementing `self.wait` at the end of each epoch is the correct behavior.

You have access to the epoch counter `epoch`, if you just want to skip the first epoch you can? E.g. if it's the first epoch, then set the best score / weights and continue.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR -- please add a unit test to check correctness and demonstrate the purpose of the change. It should be a test of a valid use case that would fail without this change.

File: keras/callbacks.py (hunk 2 above)
Comment (DachuanZhao) [reply keras/callbacks.py]: You're right. I've modified it.

**Review by fchollet — APPROVED** [review]
LGTM, thanks

**Comment by harupy** [comment]
@fchollet Is there any plan to make a patch release containing this fix?

**Comment by rchao** [comment]
Hello @harupy, the next release is estimated to be a while from now. Would `pip install keras-nightly` work if you'd like the latest change to be picked up?

**Pull request merged**